src/main/java/persistence/entity/PersistenceContextImpl.java

```diff
@@ -5,40 +5,85 @@
 
     public PersistenceContextImpl() {
         this.cacheStorage = new HashMap<>();
+        this.snapshotStorage = new HashMap<>();
     }
 
-    @Override
-    public Object get(Class<?> clazz, Long id) {
-        Map<Long, Object> entityMap = this.cacheStorage.get(clazz);
-        if (entityMap == null) {
-            return null;
+    private Map<Long, Object> getOrCreateEntityMap(Class<?> clazz) {
+        if (!this.cacheStorage.containsKey(clazz)) {
+            this.cacheStorage.put(clazz, new HashMap<>());
+        }
+        return this.cacheStorage.get(clazz);
+    }
+
+    private Map<Long, EntitySnapshot> getOrCreateSnapshotMap(Class<?> clazz) {
+        if (!this.snapshotStorage.containsKey(clazz)) {
+            this.snapshotStorage.put(clazz, new HashMap<>());
         }
+        return this.snapshotStorage.get(clazz);
+    }
+
+    @Override
+    public Object getEntity(Class<?> clazz, Long id) {
+        Map<Long, Object> entityMap = getOrCreateEntityMap(clazz);
         return entityMap.get(id);
     }
 
     @Override
-    public void put(Object entity) {
+    public void putEntity(Object entity) {
         Class<?> clazz = entity.getClass();
-        Map<Long, Object> entityMap;
-        if (this.cacheStorage.containsKey(clazz)) {
-            entityMap = this.cacheStorage.get(clazz);
-        } else {
-            entityMap = new HashMap<>();
-            this.cacheStorage.put(clazz, entityMap);
-        }
-        entityMap.put(EntityUtils.getIdValue(entity), entity);
+        Long idValue = EntityUtils.getIdValue(entity);
+
+        Map<Long, Object> entityMap = getOrCreateEntityMap(clazz);
+        entityMap.put(idValue, entity);
+
+        Map<Long, EntitySnapshot> snapshotMap = getOrCreateSnapshotMap(clazz);
+        snapshotMap.put(idValue, EntitySnapshot.of(entity));
     }
 
     @Override
-    public void remove(Object entity) {
+    public void removeEntity(Object entity) {
         Class<?> clazz = entity.getClass();
-        if (!this.cacheStorage.containsKey(clazz)) {
-            throw new IllegalArgumentException("Unable to remove such entity!");
-        }
-        Map<Long, Object> entityMap = this.cacheStorage.get(clazz);
+        Map<Long, Object> entityMap = getOrCreateEntityMap(clazz);
         entityMap.remove(EntityUtils.getIdValue(entity));
     }
+
+    @Override
+    public Map<Class<?>, Map<Long, Object>> getChangedEntities() {
+        Map<Class<?>, Map<Long, Object>> cacheStates = new HashMap<>();
+
+        for (Map.Entry<Class<?>, Map<Long, EntitySnapshot>> snapshotMapEntry : this.snapshotStorage.entrySet()) {
+            Map<Long, Object> entityMap = getOrCreateEntityMap(snapshotMapEntry.getKey());
+            Map<Long, EntitySnapshot> snapshotMap = this.snapshotStorage.get(snapshotMapEntry.getKey());
+            cacheStates.put(snapshotMapEntry.getKey(), getChangedEntitiesInType(entityMap, snapshotMap));
+        }
+        return cacheStates;
```

Comment:
```suggestion
        return snapshotStorage.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> {
            Map<Long, Object> entityMap = getOrCreateEntityMap(entry.getKey());
            Map<Long, EntitySnapshot> snapshotMap = snapshotStorage.get(entry.getKey());
            return getChangedEntitiesInType(entityMap, snapshotMap);
        }));
```
함수형 패러다임으로도 작성 가능해 보이네요